datalad/__init__.py

```diff
@@ -119,8 +119,13 @@ def set_envvar(v, val):
     consts.DATASETS_TOPURL = 'http://datasets-tests.datalad.org/'
     set_envvar('DATALAD_DATASETS_TOPURL', consts.DATASETS_TOPURL)
 
-    from datalad.tests.utils import DEFAULT_BRANCH
-    set_envvar("GIT_CONFIG_PARAMETERS", "'init.defaultBranch={}'".format(DEFAULT_BRANCH))
+    from datalad.tests.utils import (
+        DEFAULT_BRANCH,
+        DEFAULT_REMOTE,
+    )
+    set_envvar("GIT_CONFIG_PARAMETERS",
+               "'init.defaultBranch={}' 'clone.defaultRemoteName={}'"
+               .format(DEFAULT_BRANCH, DEFAULT_REMOTE))
```

Comment:
First I thought to ask: Wouldn't us "masking" default behavior of the system here with hardcoded value make tests possibly succeed in such "regulated" environment but then still fail "in real life" where no GIT_CONFIG_PARAMETERS are set?  

Then I remembered that we are making our own fake HOME for testing (below) but looking there I recalled that we do have a special case (initially introduced by debian package build procedure) where we might avoid making sanitized HOME and take `GIT_HOME` as the one to be used.

So I still wonder, if instead of hardcoding clone name and always passing them via GIT_CONFIG_PARAMETERS, shouldn't we just figure out the defaultRemoteName (and possibly defaultBranch) by running a sample invocation (in `tests.utils` and thus taking into account possible user configuration or git behavior difference and otherwise in general not pass it via `GIT_CONFIG_PARAMETERS`?

This should then run code in an environment closer to "real life", not in a such regulated one with GIT_CONFIG_PARAMETERS always being defined, but still be reflective of default git behavior.

Reply:
What failures does it risk masking?  Here's what I can think of:

  * Some DataLad logic hard codes that the default remote name is "dl-test-remote".  Or, more generally, somehow they assume some feature of "dl-test-remote" that isn't true for the set of all possible remote names.

    But (1) it seems very unlikely that any developer would make that mistake, (2) using what's on the system really just means sampling a few more possible remote names in an unsystematic way (though mostly it's just whatever is set in the temporary `HOME` since most callers are not setting `GIT_HOME`), and (3) until our minimum Git 2.30 or later, the above issue would be caught by the minimum Git CI run.

  * DataLad code is introduced that assumes that `clone.defaultRemoteName` is set, failing to provide an appropriate fallback to "origin".  This wouldn't be caught because we set it in `GIT_CONFIG_PARAMETERS`.

    This isn't the true for any spots in the current code base.  I don't think it's likely that new code will do this, but it doesn't seem too far-fetched either.

So, we could instead probe `clone.defaultRemoteName`.  We'd have to make sure we do it after the test `HOME`, if any, is set up.  Presumably we'd want to set `clone.defaultRemoteName` in the test `HOME` so that we guard against errors like the ones fixed in this PR.  That'd mean we'd catch the above error only when `GIT_HOME` is used and `clone.defaultRemoteName` isn't set there.  (That's not true for any of the CI builds currently and is unlikely to be true when a developer is testing locally.)

Weighing the above, I prefer the direct approach of this PR: explicitly configure `clone.defaultRemoteName` to something that we know would fail if DataLad doesn't handle values other than the default "origin".

But, if others are concerned about the `GIT_CONFIG_PARAMETERS` overrides, I can work on moving away from using `GIT_CONFIG_PARAMETERS` to set `clone.defaultRemoteName` and `init.defaultBranch`

Reply:
Thank you for analysis of the possible triggers of failure.  I was thinking more along the lines of: git introduces another way (env var) to allow configuration, or may be makes it possible for a remote to provide the name (may be still allowing to be overridden by a local `clone.defaultRemoteName` setting) or something along those lines.  That is why I thought that us always specifying it (while testing) could mask "natural" behavior of git if it was not specified (git still allows for it to be omitted right?).  That is why I would have preferred to not override it at all neither in GIT_CONFIG_PARAMETERS nor in git config of the tests HOME, and rather deduce default name git uses for the purpose of tests on a sample run (lazily done once).
But I do not feel strong enough to "demand" that, so we could proceed as is if nobody else has objections.

Reply:
> (git still allows for it to be omitted right?).

Yes, that is the second bullet point above.  If we were to code things in a way that assumes a value for `clone.defaultRemoteName`, that would be an error on our end.

Reply:
I said:
> So, we could instead probe `clone.defaultRemoteName`. We'd have to make sure we do it after the test `HOME`, if any, is set up.

To expand a bit on this: even if we decide that it's okay for `tests.utils` to pay the cost of probing at import time, that would happen too soon (before the temporary home is configured).  For example, this patch on top of the tip of this PR (d69c1d8da0) wouldn't capture the correct values:

<details>
<summary>diff</summary>

```diff
diff --git a/datalad/__init__.py b/datalad/__init__.py
index f99cdcfce2..9f74f309ff 100644
--- a/datalad/__init__.py
+++ b/datalad/__init__.py
@@ -119,14 +119,6 @@ def set_envvar(v, val):
     consts.DATASETS_TOPURL = 'http://datasets-tests.datalad.org/'
     set_envvar('DATALAD_DATASETS_TOPURL', consts.DATASETS_TOPURL)
 
-    from datalad.tests.utils import (
-        DEFAULT_BRANCH,
-        DEFAULT_REMOTE,
-    )
-    set_envvar("GIT_CONFIG_PARAMETERS",
-               "'init.defaultBranch={}' 'clone.defaultRemoteName={}'"
-               .format(DEFAULT_BRANCH, DEFAULT_REMOTE))
-
     # To overcome pybuild overriding HOME but us possibly wanting our
     # own HOME where we pre-setup git for testing (name, email)
     if 'GIT_HOME' in os.environ:
@@ -144,6 +136,10 @@ def set_envvar(v, val):
             os.makedirs(new_home)
         with open(os.path.join(new_home, '.gitconfig'), 'w') as f:
             f.write("""\
+[clone]
+	defaultRemoteName = dl-test-remote
+[init]
+	defaultBranch = dl-test-branch
 [user]
 	name = DataLad Tester
 	email = test@example.com
diff --git a/datalad/tests/utils.py b/datalad/tests/utils.py
index 7919f14780..9378d49242 100644
--- a/datalad/tests/utils.py
+++ b/datalad/tests/utils.py
@@ -109,20 +109,6 @@
 # Additional indicators
 on_travis = bool(os.environ.get('TRAVIS', False))
 
-if external_versions["cmd:git"] >= "2.28":
-    # The specific value here doesn't matter, but it should not be the default
-    # from any Git version to test that we work with custom values.
-    DEFAULT_BRANCH = "dl-test-branch"  # Set by setup_package().
-else:
-    DEFAULT_BRANCH = "master"
-
-if external_versions["cmd:git"] >= "2.30.0":
-    # The specific value here doesn't matter, but it should not be the default
-    # from any Git version to test that we work with custom values.
-    DEFAULT_REMOTE = "dl-test-remote"  # Set by setup_package().
-else:
-    DEFAULT_REMOTE = "origin"
-
 # additional shortcuts
 neq_ = assert_not_equal
 nok_ = assert_false
@@ -2026,6 +2012,21 @@ def get_ssh_port(host):
     return port
 
 
+@with_tempfile
+def get_default_branch_and_remote(path):
+    path = Path(path)
+    gr_a = GitRepo(path / "a")
+    branch = gr_a.call_git(["symbolic-ref", "--short", "HEAD"]).strip()
+    # Make commit to avoid warning about cloning an empty repo.
+    gr_a.call_git(["commit", "--quiet", "--allow-empty", "-mc0"])
+    gr_b = gr_a.clone(gr_a.path, str(path / "b"),
+                      clone_options={"quiet": True})
+    remotes = gr_b.get_remotes()
+    return branch, remotes[0]
+
+
+DEFAULT_BRANCH, DEFAULT_REMOTE = get_default_branch_and_remote()
+
 #
 # Context Managers
 #
```

</details>

The same is true if we move `get_default_branch_and_remote` to a separate module.  Any top-level test import would probe too soon, I think.

The main alternative that I can come up with would be changing all the test functions that use `DEFAULT_BRANCH` `DEFAULT_REMOTE` to instead call a helper function that returns those values (cached, so that the probe only happens once).

Reply:
you can just do that in the same `setup_package` fixture where we setup temp home etc.  Indeed it would be adding an unpleasant overhead of doing sensing even if a test does not care about git
(as long as we have those `DEFAULT_*`s aren't some lazy properties but IIRC there is no "properties" at module level, and there are no `@classproperty` or alike right?) 

But as a typical test in datalad does deal with git one way or another, imho it is ok.

Reply:
> you can just do that in the same `setup_package` fixture where we setup temp home etc.

True, you could mark it as a global there and have the tests import it.  But _each_ test that uses it would need to import it, so that doesn't really end up in a much better place than invoking the helper in each test.

My view is that the undefined "git may change" concern isn't worth the complication.

Reply:
just for clarity: I meant that `setup_package` would just run the check ones and modify `datalad.tests.utils.DEFAULT_*` consts as it already does e.g. for `consts.DATASETS_TOPURL` etc: https://github.com/datalad/datalad/blob/master/datalad/__init__.py#L120 and we remove the setting of `GIT_CONFIG_PARAMETERS` in the `setup_package`: this way tests would just "consult" the const as on what remote name to expect.

Reply:
> this way tests would just "consult" the const as on what remote name to expect.

As far as I can see, the general picture stays the same: tests can't do top-level imports and need to either delay them or use a helper (just like  `get_datasets_topdir` for `consts.DATASETS_TOPURL`)

Reply:
> and probably it is not worth moving those consts as GitRepo attributes, right?

I'm not strongly against this (or really any of the options being discussed), but the reason I didn't go this route is that `DEFAULT_BRANCH` or `DEFAULT_REMOTE` are used only in test code.

Reply:
oh well, since we both feel not too strongly, let's just proceed and will RF if we hit the first use case where masking would be detrimental or some expressed negative effect on DX with the growing length of each `git` invocation.

Reply:
> some expressed negative effect on DX with the growing length of each `git` invocation

Hmm, I don't understand.  It's set via an environment variable.

Reply:
ah, sorry... I thought it is handled at our level and then channeled via `git -c`s - forgot that it is a standard GIT envvar.
Well - then DX of screaming WTF while trying to execute "exactly" the same `git` invocation outside of DataLad and obtaining different results (due to `GIT_CONFIG_PARAMETERS` being set by us) ;-)